backend/GameManager.py

```diff
@@ -58,5 +82,173 @@ def get_pieces_left(self):
         }
 
     def end_turn(self):
-        self.turn = Turn.swap_turn(self.turn)
+        self.turn = Color.swap_turn(self.turn)
         return 1
+
+    def is_adjacent(self, current_row, current_col, target_row, target_col):
+        """ Check if the target position is adjacent to the current position """
+        return (target_row, target_col) in self.board.adjacent_positions_map.get((current_row, current_col), [])
+
+    def is_empty(self, row, col):
+        """ Check if the cell at (row, col) is empty """
+        return self.board.board[row][col] is None
+
+    def is_adjacent_and_empty(self, current_row, current_col, target_row, target_col):
+        """ Check if the target cell is adjacent to the current cell and is empty """
+        if self.is_adjacent(current_row, current_col, target_row, target_col):
+            if (target_row, target_col) in self.open_moves:
+                return True
+        return False
+
+    def find_empty_adjacents(self, row, col):
+        empty_adjacent_positions = []
+        all_adjacent_positions = self.board.adjacent_positions_map[(row, col)]
+
+        for row, col in all_adjacent_positions:
+            position_state = self.board.check_position(row, col)
+            if position_state == Color.EMPTY or position_state == Color.VOID:
+                empty_adjacent_positions.append((row, col))
+
+        return empty_adjacent_positions
+
+    def can_fly(self, player):
+        can_fly = False
+        if player == Color.WHITE and len(self.pieces.white_pieces) <= 3:
+            can_fly = True
+        elif player == Color.BLACK and len(self.pieces.black_pieces) <= 3:
+            can_fly = True
+        return can_fly
+
+    '''
+    the select_piece func should return the available empty adjacent positions to where the piece can be moved 
+    or throw exception if it should be selectable
+    '''
+    def select_piece(self, row, col):
+        if not self.placement_complete():
+            raise Exception("SelectionError -- Cannot select pieces during placement phase")
+
+        cell = self.board.get_cell(row, col)
+        self.selected_piece = (row, col)
+
+        if self.turn == cell.get_state():
+            # calculate the available adjacent positions
+            if self.can_fly(self.turn):
+                # if can fly, return all open positions
+                return self.open_moves
+            else:
+                # if more than 3 pieces left, user can only move to empty adjacent positions
+                return self.find_empty_adjacents(row, col)
+
+        self.selected_piece = None
+        raise Exception("SelectionError -- Invalid piece selection")
+
+
+    def move_piece(self, target_row, target_col):
+        if not self.selected_piece:
+            Exception("MoveError -- No piece selected")
```

Comment:
Exceptions have to have `raise` before them so that they stop the rest of the function from running. I know I haven't pushed the acceptance tests yet, but this doesn't pass them because the exception doesn't get raised.

Reply:
I'm not going to mark the rest of the unraised Exceptions unless you want me to.

Reply:
i think may create our custom exception and raise in the files so for now may be you can modify that test or idk comment it.